src/components/select/select.jsx

```diff
@@ -0,0 +1,231 @@
+    icon={props.readOnly ? iconCaretDownDisabled : iconCaretDown}
+    label={props.readOnly ? 'Select is not editable' : 'Open choices listbox'}
+  />);
+
+  const defaultValue = value ? props.displayValueEvaluator(value || props.value) : '';
+
+  const wrapperRef = useRef(null);
+  const handleDropdownClose = () => {
+    setShowOptions(false);
+    setSearchValue(defaultValue);
+  };
+  useDropdownClose(wrapperRef, showOptions, handleDropdownClose);
+
+  const clear = () => {
+    setValue(undefined);
+    setSearchValue(undefined);
+    inputRef.current.focus();
+  };
+
+  const clearIcon = () => {
+    if (!props.readOnly && (value || searchValue)) {
+      return (<IconButton
+        className={styles['clear-button']}
+        icon={iconClear}
+        label={'Remove selected choice'}
+        onPress={clear}
+      />);
+    }
+
+    return undefined;
+  };
+
+  function onClick() {
+    if (!props.readOnly) setShowOptions(true);
+  }
+
+  function onKeyDown(event) {
+    switch (event.key) {
+      case 'Enter':
+        event.preventDefault();
+        if (!showOptions && !props.readOnly) setShowOptions(true);
+        break;
+      case 'Escape':
+        event.preventDefault();
+        setShowOptions(false);
+        break;
+      default:
+    }
+  }
+
+  function updateOptionVisibility() {
+    if (searchValue) {
+      props.listOptionRefs.forEach((listOptionRef) => {
+        if (listOptionRef.current) {
+          const listOptionDisplayValue = props.displayValueEvaluator(listOptionRef.current.value);
+
+          listOptionRef.current.setVisible(listOptionDisplayValue.includes(searchValue));
+        }
+      });
+    } else {
+      props.listOptionRefs.forEach((listOptionRef) => {
+        if (listOptionRef.current) {
+          listOptionRef.current.setVisible(true);
+        }
+      });
+    }
+  }
+
+  function updateOptionSelected() {
+    props.listOptionRefs.forEach((listOptionRef) => {
+      if (listOptionRef.current) {
+        listOptionRef.current.setSelected(JSON.stringify(value) === JSON.stringify(listOptionRef.current.value));
+      }
+    });
+  }
+
+  function onSelectionChange(event) {
+    const selectedValue = event.target.value;
+    setValue(selectedValue);
+    setSearchValue(selectedValue.label);
+    setShowOptions(false);
+    inputRef.current.focus();
+  }
+
+  function onSearchChange(event) {
+    const newValue = event.target.value;
+
+    if (newValue === '') {
+      clear();
+      return;
+    }
+
+    setSearchValue(newValue);
+    setShowOptions(true);
+  }
+
+  useImperativeHandle(selfRef, () => ({
+    get dirty() {
+      return props.value !== this.value;
+    },
+    id: props.id,
+    name: props.name,
+    get value() {
+      return value;
+    },
+  }));
+
+  useEffect(() => {
+    props.onChange({ target: selfRef.current });
+  }, [value]);
+
+  useEffect(() => {
+    updateOptionVisibility();
+  }, [searchValue, props.listOptionRefs]);
+
+  useEffect(() => {
```

Comment:
Looking at this file, there is still a feeling of "woah, there's a lot going on here" -- and I think there isn't a way around it. But, I think an idea that might help:

Looks like we only use `updateOptionSelected` in this `useEffect`. We can just inline the function. Something like:

```suggestion
  useEffect(function updateOptionSelected() {
    ...
  }, [value, showOptions]);
```

Reply:
If we like the idea, let's do the same for the other useEffects.

Reply:
I'll try this fo sho